[2.6] NEXUS-5704: 500 Internal Server Error when "If-None-Match" in header

cstamas opened the pull request:
Restlet 1.1 is very picky (and have a bug) how it handles ETags: it requires "by the book" how HTTP request formats ETags, they must be quoted. In other case, the header presence is detected (condition list is non-empty), but unexpected `null` value is returned for tag value (the condition list will have one element, which is `null`). This causes NPE in Nexus code.

"Vanilla" Nexus (even without this change) _works properly_ for HTTP conditional GETs using `if-modified-since` headers, as shown in the linked issue. Still, with unquoted ETag NPE does happens in Nexus codebase, and this pull fixes it.

This pull merely fixes that single point of NPE by adding nullcheck, and also simplifies the `renderStorageFileItem()` method, but DOES NOT fix the Restlet library in this manner. Nexus will still NPE if HTTP request carries badly formatted ETag. After this change, badly formatted ETags will continue to cause 500 Internal Server Error responses, but this time deeply in Restlet code (where also Tag inspection is tried, without nullchecks around it).

Relates to issue
https://issues.sonatype.org/browse/NEXUS-5704

cstamas committed:
NEXUS-5704: NPE when client uses if-none-match.

This fixes the NPE in Nexus code, but still REstlet itself
is NPE-prone if the tag is not in quotes as it seems.

cstamas committed:
NEXUS-5704: Reworked NPE protection

And simplified method

ifedorenko commented:
Is it hard to unit test this change? And I do mean "unit test", I don't believe full-blown integration test is a good idea in this case.

cstamas commented:
I believe we have UTs for some resources, will take take a peek into those, and how hard would be to cover this change...

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5704-if-none-match

cstamas committed:
NEXUS-5704: UT for the two changed spots.

Added unit tests to verify there is no more NPE
in the ResourceStoreRequest creation when
if-modified-since header is used.

Another spot tested is verifying that
file item does honor and behaves well as regarding
the handling of request carried if-modified-since
conditions.

cstamas commented:
Added UT that verifies the two spots changed.

cstamas commented:
CI
https://bamboo.zion.sonatype.com/browse/NXO-OSSF25-1

kellyrob99 commented on plugins/restlet1x/nexus-restlet1x-plugin/src/test/java/org/sonatype/nexus/rest/ResourceStoreContentPlexusResourceTest.java:
```diff
@@ -214,4 +239,111 @@ public void testNexus5155OmitCacheHeadersForLinkedFileItems()
+            when( attributes.get( StorageFileItem.DIGEST_SHA1_KEY ) ).thenReturn( "1234567890" );
+
+            underTest.renderStorageFileItem( request, fileItem );
+            assertThat( "ResourceException was expected to be thrown", false );
```
The intent of the test would probably be clearer here if you just used fail() with the same message

kellyrob99 commented:
So IIUC the particular user that reported this issue would still have to properly format their Etag in order to use it successfully, but at least now they won't get an NPE?

cstamas commented:
CI +1'd
https://bamboo.zion.sonatype.com/browse/NXO-OSSF25-1

cstamas commented:
@kellyrob99 Will change the UT to make it more clearer.

And yes, Restlet framework suffers from "unquoted ETag" bug, but after this fix the nasty NPE will not happen in Nexus codebase, but even with this, users will still have to "properly" format ETags.

cstamas committed:
NEXUS-5704: cleanup and commented changes

Using fail() instead to make intention clear.
Cleanup of imports.

adreghiciu commented:
+1

cstamas commented:
Moved to collapsed project as https://github.com/sonatype/nexus-oss/pull/8
